src/raid.cpp

```diff
@@ -0,0 +1,911 @@
+/**
+ * @file mega/raid.cpp
+ * @brief helper classes for managing cloudraid downloads
+ *
+ * (c) 2013-2019 by Mega Limited, Auckland, New Zealand
+ *
+ * This file is part of the MEGA SDK - Client Access Engine.
+ *
+ * Applications using the MEGA API must present a valid application key
+ * and comply with the the rules set forth in the Terms of Service.
+ *
+ * The MEGA SDK is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * @copyright Simplified (2-clause) BSD License.
+ *
+ * You should have received a copy of the license along with this
+ * program.
+ */
+
+#include "mega/raid.h"
+
+#include "mega/transfer.h" 
+#include "mega/testhooks.h" 
+#include "mega.h" // for thread definitions
+
+namespace mega
+{
+struct FaultyServers
+{
+    // Records URLs that had recent problems, so we can start the next raid download with URLs that can work first try.
+    // In particular this is useful when one server in a raid set is unavailable for an extended period
+    // This class may be shared amongst many megaclients, so thread safety is needed
+    typedef map<string, m_time_t> Map;
+    Map recentFails;
+    MUTEX_CLASS m;
+
+    FaultyServers()
+    {
+        m.init(false);
+    }
+
+    string server(const string& url)
+    {
+        size_t n = url.find("://");
+        if (n != string::npos)
+        {
+            n += 3;
+            size_t m = url.find("/", n);
+            if (m != string::npos)
+            {
+                return url.substr(n, m - n);
+            }
+        }
+        return "";
+    }
+
+    void add(const string& url)
+    {
+        MutexGuard g(m);
+        recentFails[server(url)] = m_time();
+    }
+
+    unsigned selectWorstServer(vector<string> urls)
+    {
+        // start with 6 connections and drop the slowest to respond
+        // (unless we know one of the 6 URLs is on a dodgy server, in which case start with the other 5)
+       // unsigned worstindex = RAIDPARTS;
+        unsigned worstindex = rand() % RAIDPARTS;
+
+        MutexGuard g(m);
+        if (!recentFails.empty())
+        {
+            m_time_t now = m_time();
+            m_time_t worsttime = now - 10 * 3600;
```

Comment:
I suggest to use `NEVER`. Not sure if possible, but if any of the `url`s received as input is a hit in `recentFails` and the `ts` associated is older than 10 hours (likely to happen in MEGAsync), it would be excluded from the search of the worst index.
```suggestion
            m_time_t worsttime = NEVER;
```

Reply:
I was thinking that if a server is down, or temporarily can't serve that file, etc, it will probably be fixed on some timescale, maybe within a few hours and then we should resume normal operation rather than having the program avoid that server forever (in the absence of any other server getting into the data structure).   10 hours is probably not too frequent to try and see if problems have been resolved, and not too slow - what do you think?  thanks

Reply:
It's a good idea, indeed, to avoid using a faulty server for 10h and after that period of time to give it another chance. However, in that case, I think the:
```if (j != recentFails.end() && j->second > worsttime) ```
should be:
```if (j != recentFails.end() && j->second <= worsttime) ```
in order to select the faulty URL as the worst one, unless more than 10h have already elapsed. After 10h, the `worstindex` may randomly be assigned to the faulty one again.

Reply:
This is the commit with the comments: https://github.com/meganz/sdk/commit/bc999f5d68ad34944cd2a681982d1bd4c786cc97